Update to net.mikehardy:google-analytics-java:2.0.11

There are issues due to transitive dependency updates after the analytics update - reverting for now, let's move it to a separate issue.

```
> Task :AnkiDroid:compileDebugJavaWithJavac
C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\anki\web\HttpFetcher.java:126: warning: [deprecation] create(MediaType,byte[]) in RequestBody has been deprecated
                requestBuilder.post(RequestBody.create(null, new byte[0]));
                                               ^
error: warnings found and -Werror specified
warnings found and -Werror specified

C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\libanki\sync\CountingFileRequestBody.java:64: warning: [deprecation] buffer() in BufferedSink has been deprecated
            while ((read = source.read(sink.buffer(), SEGMENT_SIZE)) != -1) {
                                           ^
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error
2 warnings

> Task :AnkiDroid:compileDebugJavaWithJavac FAILED
```

_Originally posted by @david-allison-1 in https://github.com/ankidroid/Anki-Android/pull/7055#issuecomment-690363780_

I took another try at this and I was able to fix all the undeclared transitive usages of the java8 objects imported by the java7-compatible flavor of google-analytics-java

I still haven't found a super-easy way to fix this though:

```
com.ichi2.anki.FieldEditLineTest > testSaveRestore[TestingAVD_API-21_TAG-GAPI(AVD) - 5.0.2] FAILED 
        java.lang.NoClassDefFoundError: Failed resolution of: Ljava/util/concurrent/CompletableFuture;
        at com.brsanthu.googleanalytics.internal.GoogleAnalyticsImpl.postAsync(GoogleAnalyticsImpl.java:123)
[InstrumentationResultParser]: Ignoring repeated failed event for com.ichi2.anki.FieldEditLineTest#testSaveRestore. Stack: java.lang.AssertionError: Activity never becomes requested state "[CREATED, STARTED, RESUMED, DESTROYED]" (last lifecycle transition = "STARTED")
``` 

That object is not included in the java8 de-sugaring, so this is effectively on hold until I either find an easy way to package that dependency up *in* the google-analytics-java artifact (not likely, it's work that has little value) or we go to minSdkVersion >= 24.

--- a/AnkiDroid/build.gradle
+++ b/AnkiDroid/build.gradle
@@ -390,6 +390,8 @@ dependencies {
     implementation project(":libanki")
     implementation project(":vbpd")
 
+    implementation(libs.google.analytics.kt)
+
     implementation libs.androidx.activity
     implementation libs.androidx.annotation
     implementation libs.androidx.appcompat

--- a/AnkiDroid/build.gradle
+++ b/AnkiDroid/build.gradle
@@ -99,6 +99,15 @@ android {
         testApplicationId = "com.ichi2.anki.tests"
         vectorDrawables.useSupportLibrary = true
         testInstrumentationRunner = 'com.ichi2.testutils.NewCollectionPathTestRunner'
+
+        Properties properties = new Properties()
+        def localPropsFile = project.rootProject.file("local.properties")
+        if (localPropsFile.exists()) {
+            properties.load(localPropsFile.newDataInputStream())
+        }
+
+        def apiKey = properties.getProperty("ANALYTICS_API_KEY") ?: System.getenv("ANALYTICS_API_KEY") ?: "DUMMY_API_XXX"
+        buildConfigField "String", "ANALYTICS_API_KEY", "\"${apiKey}\""
     }
     signingConfigs {
         release {

--- a/AnkiDroid/build.gradle
+++ b/AnkiDroid/build.gradle
@@ -409,6 +409,8 @@ dependencies {
     implementation project(":libanki")
     implementation project(":vbpd")
 
+    implementation(libs.google.analytics.kt)
+
     implementation libs.androidx.activity
     implementation libs.androidx.annotation
     implementation libs.androidx.appcompat

--- a/AnkiDroid/build.gradle
+++ b/AnkiDroid/build.gradle
@@ -110,6 +110,22 @@ android {
         testApplicationId = "com.ichi2.anki.tests"
         vectorDrawables.useSupportLibrary = true
         testInstrumentationRunner = 'com.ichi2.testutils.NewCollectionPathTestRunner'
+
+        // GA4 Measurement Protocol api_secret. Sourced from `local.properties`
+        // (ANALYTICS_API_KEY) or the ANALYTICS_API_KEY env var; falls back to
+        // a dummy value so contributor builds still compile.
+        //
+        // The dummy key is rejected by GA's ingest, so analytics requests from
+        // unsigned/local builds go nowhere this means `setDevMode()` (100%
+        // sampling) has no observable effect on a build without the real key.
+        Properties properties = new Properties()
+        def localPropsFile = project.rootProject.file("local.properties")
+        if (localPropsFile.exists()) {
+            properties.load(localPropsFile.newDataInputStream())
+        }
+
+        def apiKey = properties.getProperty("ANALYTICS_API_KEY") ?: System.getenv("ANALYTICS_API_KEY") ?: "DUMMY_API_XXX"
+        buildConfigField "String", "ANALYTICS_API_KEY", "\"${apiKey}\""
     }
     signingConfigs {
         release {

--- a/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnalyticsExceptionHandler.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnalyticsExceptionHandler.kt
@@ -1,0 +1,73 @@
+/*
+ * Copyright (c) 2026 Ashish Yadav <mailtoashish693@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation; either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki.analytics
+
+import timber.log.Timber
+
+/**
+ * A custom [Thread.UncaughtExceptionHandler] designed to intercept fatal app crashes,
+ * log them to the analytics provider (e.g., GA4), and safely pass them down the chain
+ * to the original exception handler.
+ *
+ * @property originalHandler The pre-existing exception handler in the thread's chain.
+ * This is called immediately after the analytics report is dispatched.
+ * @property sendExceptionReport A delegated function responsible for transmitting the
+ * [Throwable] to the analytics engine. The boolean parameter
+ * indicates if the exception is fatal (always `true` here).
+ */
+class AnalyticsExceptionHandler(
+    val originalHandler: Thread.UncaughtExceptionHandler?,
+    private val sendExceptionReport: (Throwable, Boolean) -> Unit,
+) : Thread.UncaughtExceptionHandler {
+    override fun uncaughtException(
+        thread: Thread,
+        throwable: Throwable,
+    ) {
+        sendExceptionReport(throwable, true)
+        originalHandler?.uncaughtException(thread, throwable)
+    }
+
+    companion object {
+        /**
+         * Safely installs the [AnalyticsExceptionHandler] as the default handler for the thread.
+         *
+         * @param sendExceptionReport The function reference to be used for sending the crash data.
+         */
+        fun install(sendExceptionReport: (Throwable, Boolean) -> Unit) {
+            val currentHandler = Thread.getDefaultUncaughtExceptionHandler()
+            if (currentHandler !is AnalyticsExceptionHandler) {
+                Thread.setDefaultUncaughtExceptionHandler(
+                    AnalyticsExceptionHandler(currentHandler, sendExceptionReport),
+                )
+                Timber.d("AnalyticsExceptionHandler installed.")
+            }
+        }
+
+        /**
+         * Uninstalls the [AnalyticsExceptionHandler] by restoring the thread's default
+         * exception handler back to the [originalHandler] that was present before installation.
+         */
+        fun uninstall() {
+            val currentHandler = Thread.getDefaultUncaughtExceptionHandler()
+            if (currentHandler is AnalyticsExceptionHandler) {
+                Thread.setDefaultUncaughtExceptionHandler(currentHandler.originalHandler)
+                Timber.d("AnalyticsExceptionHandler uninstalled.")
+            }
+        }
+    }
+}

--- a/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnalyticsExceptionHandler.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnalyticsExceptionHandler.kt
@@ -1,0 +1,59 @@
+// SPDX-FileCopyrightText: 2026 Ashish Yadav <mailtoashish693@gmail.com>
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+package com.ichi2.anki.analytics
+
+import timber.log.Timber
+
+/**
+ * A custom [Thread.UncaughtExceptionHandler] designed to intercept fatal app crashes,
+ * log them to the analytics provider (e.g., GA4), and safely pass them down the chain
+ * to the original exception handler.
+ *
+ * @property originalHandler The pre-existing exception handler in the thread's chain.
+ * This is called immediately after the analytics report is dispatched.
+ * @property sendExceptionReport A delegated function responsible for transmitting the
+ * [Throwable] to the analytics engine. The boolean parameter
+ * indicates if the exception is fatal (always `true` here).
+ */
+class AnalyticsExceptionHandler(
+    val originalHandler: Thread.UncaughtExceptionHandler?,
+    private val sendExceptionReport: (Throwable, Boolean) -> Unit,
+) : Thread.UncaughtExceptionHandler {
+    override fun uncaughtException(
+        thread: Thread,
+        throwable: Throwable,
+    ) {
+        sendExceptionReport(throwable, true)
+        originalHandler?.uncaughtException(thread, throwable)
+    }
+
+    companion object {
+        /**
+         * Safely installs the [AnalyticsExceptionHandler] as the default handler for the thread.
+         *
+         * @param sendExceptionReport The function reference to be used for sending the crash data.
+         */
+        fun install(sendExceptionReport: (Throwable, Boolean) -> Unit) {
+            val currentHandler = Thread.getDefaultUncaughtExceptionHandler()
+            if (currentHandler !is AnalyticsExceptionHandler) {
+                Thread.setDefaultUncaughtExceptionHandler(
+                    AnalyticsExceptionHandler(currentHandler, sendExceptionReport),
+                )
+                Timber.d("AnalyticsExceptionHandler installed.")
+            }
+        }
+
+        /**
+         * Uninstalls the [AnalyticsExceptionHandler] by restoring the thread's default
+         * exception handler back to the [originalHandler] that was present before installation.
+         */
+        fun uninstall() {
+            val currentHandler = Thread.getDefaultUncaughtExceptionHandler()
+            if (currentHandler is AnalyticsExceptionHandler) {
+                Thread.setDefaultUncaughtExceptionHandler(currentHandler.originalHandler)
+                Timber.d("AnalyticsExceptionHandler uninstalled.")
+            }
+        }
+    }
+}

--- a/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnalyticsSamplePercentage.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnalyticsSamplePercentage.kt
@@ -1,0 +1,38 @@
+// SPDX-FileCopyrightText: 2026 Ashish Yadav <mailtoashish693@gmail.com>
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+package com.ichi2.anki.analytics
+
+/**
+ * A type-safe wrapper representing the sampling rate for Google Analytics (GA) events.
+ *
+ * This defines the percentage of events sent to GA, while the remainder are dropped
+ * client-side to limit reporting volume. It is backed by an [Int] ranging from `0..100`.
+ *
+ * The [Uninitialized] state serves as a sentinel value indicating that the rate has not
+ * yet been loaded from resources, eliminating the need for callers to handle custom
+ * out-of-band integers for this state.
+ */
+@JvmInline
+value class AnalyticsSamplePercentage(
+    val value: Int,
+) {
+    init {
+        require(value == UNINITIALIZED_VALUE || value in 0..100) {
+            "Analytics sample percentage must be in 0..100 or Uninitialized (-1), got $value"
+        }
+    }
+
+    /** `true` once the percentage has been loaded from resources or set explicitly. */
+    val isInitialized: Boolean get() = value != UNINITIALIZED_VALUE
+
+    companion object {
+        private const val UNINITIALIZED_VALUE = -1
+
+        /** Sentinel: not yet loaded from resources. */
+        val Uninitialized = AnalyticsSamplePercentage(UNINITIALIZED_VALUE)
+
+        /** 100% of events are sent; used by [AnkiDroidUsageAnalytics.setDevMode]. */
+        val Full = AnalyticsSamplePercentage(100)
+    }
+}

--- a/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnkiDroidUsageAnalytics.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnkiDroidUsageAnalytics.kt
@@ -1,0 +1,201 @@
+/*
+ * Copyright (c) 2026 Ashish Yadav <mailtoashish693@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation; either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki.analytics
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.annotation.VisibleForTesting
+import androidx.core.content.edit
+import com.criticalay.GoogleAnalytics
+import com.criticalay.GoogleAnalyticsConfig
+import com.ichi2.anki.AnkiDroidApp
+import com.ichi2.anki.BuildConfig
+import com.ichi2.anki.R
+import com.ichi2.anki.preferences.sharedPrefs
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicBoolean
+
+// TODO: write test for this class
+object AnkiDroidUsageAnalytics {
+    const val ANALYTICS_OPTIN_KEY = "analytics_opt_in"
+    private const val ANALYTICS_CLIENT_ID = "googleAnalyticsClientId"
+    private const val ANALYTICS_PREFS = "analyticsPrefs"
+
+    @Volatile private var analytics: GoogleAnalytics? = null
+
+    @Volatile private var optIn = false
+
+    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
+    private val clientId: String by lazy { getOrCreateClientId(AnkiDroidApp.instance) }
+
+    private val sharedPrefsListener =
+        SharedPreferences.OnSharedPreferenceChangeListener { prefs, key ->
+            if (key == ANALYTICS_OPTIN_KEY) {
+                optIn = prefs.getBoolean(key, false)
+                Timber.i("Setting analytics opt-in to: %b", optIn)
+            }
+        }
+
+    private var samplePercentage: Int = -1
+
+    lateinit var preferencesWhoseChangesShouldBeReported: Set<String>
+        private set
+
+    fun initialize(context: Context) {
+        val appContext = context.applicationContext
+
+        Timber.i("AnkiDroidUsageAnalytics:: initialize()")
+
+        if (analytics == null) {
+            analytics =
+                GoogleAnalytics.builder {
+                    measurementId = appContext.getString(R.string.ga_trackingId)
+                    apiSecret = BuildConfig.ANALYTICS_API_KEY
+                    appName = appContext.getString(R.string.app_name)
+                    appVersion = BuildConfig.VERSION_NAME
+                    enabled = true
+                    samplePercentage = getAnalyticsSamplePercentage(appContext)
+                    debug = false
+                }
+        }
+
+        handlePreferences(appContext)
+        initializePrefKeys(appContext)
+
+        AnalyticsExceptionHandler.install(this::sendAnalyticsException)
+    }
+
+    private fun handlePreferences(context: Context) {
+        val userPrefs = context.sharedPrefs()
+        optIn = userPrefs.getBoolean(ANALYTICS_OPTIN_KEY, false)
+        userPrefs.registerOnSharedPreferenceChangeListener(sharedPrefsListener)
+    }
+
+    fun reinitialize(context: Context) {
+        Timber.i("reInitialize()")
+        AnalyticsExceptionHandler.uninstall()
+
+        serviceScope.launch {
+            runCatching { analytics?.flush() }.onFailure { e ->
+                Timber.e(e, "Failed to flush analytics")
+            }
+            analytics = null
+            initialize(context)
+        }
+    }
+
+    fun sendAnalyticsScreenView(screen: Any) = sendAnalyticsScreenView(screen.javaClass.simpleName)
+
+    fun sendAnalyticsScreenView(screenName: String) {
+        Timber.d("AnkiDroidUsageAnalytics: screenView($screenName)")
+        if (!optIn) return
+        analytics?.screenView(clientId)?.screenName(screenName)?.sendAsync()
+    }
+
+    fun sendAnalyticsEvent(
+        category: String,
+        action: String,
+        value: Int? = null,
+        label: String? = null,
+    ) {
+        Timber.d("AnkiDroidUsageAnalytics: event(category=$category action=$action)")
+        if (!optIn) return
+        analytics
+            ?.event(clientId)
+            ?.category(category)
+            ?.action(action)
+            ?.apply { label?.let { label(it) } }
+            ?.apply { value?.let { value(it) } }
+            ?.sendAsync()
+    }
+
+    fun sendAnalyticsException(
+        t: Throwable,
+        fatal: Boolean,
+    ) {
+        val cause = getRootCause(t)
+        sendAnalyticsException("${cause::class.simpleName}: ${cause.message}", fatal)
+    }
+
+    fun sendAnalyticsException(
+        description: String,
+        fatal: Boolean,
+    ) {
+        if (!optIn) return
+
+        Timber.d("AnkiDroidUsageAnalytics: exception(fatal=$fatal)")
+        analytics
+            ?.exception(clientId)
+            ?.description(description.take(150))
+            ?.fatal(fatal)
+            ?.sendAsync()
+    }
+
+    private fun getRootCause(t: Throwable): Throwable {
+        var result = t
+        while (result.cause != null && result.cause !== result) {
+            result = result.cause!!
+        }
+        return result
+    }
+
+    private fun getOrCreateClientId(context: Context): String {
+        Timber.d("AnkiDroidUsageAnalytics:: getting client Id")
+        val prefs = context.getSharedPreferences(ANALYTICS_PREFS, Context.MODE_PRIVATE)
+        return prefs.getString(ANALYTICS_CLIENT_ID, null) ?: UUID.randomUUID().toString().also {
+            prefs.edit { putString(ANALYTICS_CLIENT_ID, it) }
+        }
+    }
+
+    private fun getAnalyticsSamplePercentage(context: Context): Int {
+        Timber.d("AnkiDroidUsageAnalytics:: getting sample percentage")
+        if (samplePercentage == -1) {
+            samplePercentage = context.resources.getInteger(R.integer.ga_sampleFrequency)
+        }
+        return samplePercentage
+    }
+
+    private fun initializePrefKeys(context: Context) {
+        Timber.d("AnkiDroidUsageAnalytics:: initializing pref keys")
+        preferencesWhoseChangesShouldBeReported =
+            AnalyticsConstants.reportablePrefKeys.mapTo(
+                HashSet(AnalyticsConstants.reportablePrefKeys.size),
+            ) { context.getString(it) }
+    }
+
+    var isEnabled: Boolean
+        get() = optIn
+        set(value) {
+
+            AnkiDroidApp.instance.sharedPrefs().edit {
+                putBoolean(ANALYTICS_OPTIN_KEY, value)
+            }
+        }
+
+    fun setDevMode(context: Context) {
+        Timber.d("setDevMode() re-configuring for development analytics tagging")
+        samplePercentage = 100
+        reinitialize(context)
+    }
+}

--- a/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnkiDroidUsageAnalytics.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnkiDroidUsageAnalytics.kt
@@ -1,0 +1,248 @@
+// SPDX-FileCopyrightText: 2026 Ashish Yadav <mailtoashish693@gmail.com>
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+package com.ichi2.anki.analytics
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.core.content.edit
+import com.criticalay.GoogleAnalytics
+import com.ichi2.anki.AnkiDroidApp
+import com.ichi2.anki.BuildConfig
+import com.ichi2.anki.R
+import com.ichi2.anki.common.annotations.NeedsTest
+import com.ichi2.anki.common.utils.ext.getRootCause
+import com.ichi2.anki.preferences.sharedPrefs
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import timber.log.Timber
+import java.util.UUID
+
+@NeedsTest("Add coverage for opt-in handling, client id persistence and event/exception sending")
+object AnkiDroidUsageAnalytics {
+    private const val ANALYTICS_OPTIN_KEY = "analytics_opt_in"
+    private const val ANALYTICS_CLIENT_ID = "googleAnalyticsClientId"
+
+    /**
+     * Hard cap on the length of an exception description sent to GA.
+     * GA's `exception_description` parameter is bounded truncate
+     * defensively so an overly long stack/message isn't rejected at the wire.
+     */
+    private const val MAX_EXCEPTION_DESCRIPTION_LENGTH = 150
+
+    /**
+     * Dedicated prefs file (separate from the user-facing app preferences) for the
+     * analytics client id. Keeping it isolated avoids exposing the id through
+     * preference screens, backups, or bulk prefs operations on the main file.
+     *
+     * The id is install-scoped (one per device install) rather than per-profile
+     * profiles share the same analytics client id, which matches GA's expectations
+     * and avoids fragmenting analytics across profile switches.
+     */
+    private const val ANALYTICS_PREFS = "analyticsPrefs"
+
+    @Volatile private var analytics: GoogleAnalytics? = null
+
+    @Volatile private var optIn = false
+
+    /**
+     * Application context captured during [initialize]. Held here so we don't
+     * rely on [AnkiDroidApp.instance] from background paths the singleton can
+     * be uninitialized in rare Android scenarios (e.g. BackupManager) and
+     * analytics is a startup concern that must not crash.
+     */
+    private lateinit var appContext: Context
+
+    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
+    private val clientId: String by lazy { getOrCreateClientId(appContext) }
+
+    private val sharedPrefsListener =
+        SharedPreferences.OnSharedPreferenceChangeListener { prefs, key ->
+            if (key == ANALYTICS_OPTIN_KEY) {
+                optIn = prefs.getBoolean(key, false)
+                Timber.i("Setting analytics opt-in to: %b", optIn)
+            }
+        }
+
+    /**
+     * Cached percentage of analytics events that are actually sent; the rest
+     * are dropped client-side to limit volume.
+     *
+     * Starts as [AnalyticsSamplePercentage.Uninitialized] and is loaded lazily
+     * from `R.integer.ga_sampleFrequency` on first read. [setDevMode] overrides
+     * it to [AnalyticsSamplePercentage.Full] so every event is sent during
+     * development.
+     */
+    private var samplePercentage: AnalyticsSamplePercentage = AnalyticsSamplePercentage.Uninitialized
+
+    /**
+     * Resolved preference keys (the localized string-resource values, not
+     * resource ids) whose changes both the key and the new value are
+     * reported to analytics. Populated during [initialize] from
+     * [AnalyticsConstants.reportablePrefKeys].
+     */
+    lateinit var reportablePreferences: Set<String>
+        private set
+
+    fun initialize(context: Context) {
+        appContext = context.applicationContext
+
+        Timber.i("AnkiDroidUsageAnalytics:: initialize()")
+
+        // Read opt-in before building the client so `enabled` reflects the
+        // user's choice rather than the default.
+        handlePreferences(appContext)
+
+        if (analytics == null) {
+            analytics =
+                GoogleAnalytics.builder {
+                    measurementId = appContext.getString(R.string.ga_trackingId)
+                    apiSecret = BuildConfig.ANALYTICS_API_KEY
+                    appName = appContext.getString(R.string.app_name)
+                    appVersion = BuildConfig.VERSION_NAME
+                    enabled = optIn
+                    samplePercentage = getAnalyticsSamplePercentage(appContext)
+                    debug = false
+                }
+        }
+
+        initializePrefKeys(appContext)
+
+        AnalyticsExceptionHandler.install(this::sendAnalyticsException)
+    }
+
+    private fun handlePreferences(context: Context) {
+        val userPrefs = context.sharedPrefs()
+        optIn = userPrefs.getBoolean(ANALYTICS_OPTIN_KEY, false)
+        userPrefs.registerOnSharedPreferenceChangeListener(sharedPrefsListener)
+    }
+
+    fun reinitialize(context: Context) {
+        Timber.i("reInitialize()")
+        AnalyticsExceptionHandler.uninstall()
+
+        serviceScope.launch {
+            runCatching { analytics?.flush() }.onFailure { e ->
+                Timber.w(e, "Failed to flush analytics")
+            }
+            analytics = null
+            initialize(context)
+        }
+    }
+
+    /**
+     * Records a screen view named after the runtime class of [screen].
+     *
+     * Uses [Class.getSimpleName] so the screen name is the unqualified class
+     * name (e.g. `DeckPicker`, not `com.ichi2.anki.DeckPicker`). Pass an
+     * activity/fragment/`this` from the screen you want to record.
+     */
+    fun sendAnalyticsScreenView(screen: Any) = sendAnalyticsScreenView(screen.javaClass.simpleName)
+
+    fun sendAnalyticsScreenView(screenName: String) {
+        Timber.d("AnkiDroidUsageAnalytics: screenView($screenName)")
+        if (!optIn) return
+        analytics?.screenView(clientId)?.screenName(screenName)?.sendAsync()
+    }
+
+    fun sendAnalyticsEvent(
+        category: String,
+        action: String,
+        value: Int? = null,
+        label: String? = null,
+    ) {
+        Timber.d("AnkiDroidUsageAnalytics: event(category=$category action=$action)")
+        if (!optIn) return
+        val analytics = analytics ?: return
+        val event =
+            analytics
+                .event(clientId)
+                .category(category)
+                .action(action)
+        label?.let { event.label(it) }
+        value?.let { event.value(it) }
+        event.sendAsync()
+    }
+
+    fun sendAnalyticsException(
+        t: Throwable,
+        fatal: Boolean,
+    ) {
+        val cause = t.getRootCause()
+        sendAnalyticsException("${cause::class.simpleName}: ${cause.message}", fatal)
+    }
+
+    fun sendAnalyticsException(
+        description: String,
+        fatal: Boolean,
+    ) {
+        if (!optIn) return
+
+        Timber.d("AnkiDroidUsageAnalytics: exception(fatal=$fatal)")
+        val analytics = analytics ?: return
+        analytics
+            .exception(clientId)
+            .description(description.take(MAX_EXCEPTION_DESCRIPTION_LENGTH))
+            .fatal(fatal)
+            .sendAsync()
+    }
+
+    private fun getOrCreateClientId(context: Context): String {
+        Timber.d("AnkiDroidUsageAnalytics:: getting client Id")
+        val prefs = context.getSharedPreferences(ANALYTICS_PREFS, Context.MODE_PRIVATE)
+        return prefs.getString(ANALYTICS_CLIENT_ID, null) ?: UUID.randomUUID().toString().also {
+            prefs.edit { putString(ANALYTICS_CLIENT_ID, it) }
+        }
+    }
+
+    private fun getAnalyticsSamplePercentage(context: Context): Int {
+        Timber.d("AnkiDroidUsageAnalytics:: getting sample percentage")
+        if (!samplePercentage.isInitialized) {
+            samplePercentage =
+                AnalyticsSamplePercentage(context.resources.getInteger(R.integer.ga_sampleFrequency))
+        }
+        return samplePercentage.value
+    }
+
+    /**
+     * Resolves [AnalyticsConstants.reportablePrefKeys] (string-resource ids)
+     * to their localized key strings via [context] and caches the result in
+     * [reportablePreferences] for fast membership checks at the call site.
+     */
+    private fun initializePrefKeys(context: Context) {
+        Timber.d("AnkiDroidUsageAnalytics:: initializing pref keys")
+        reportablePreferences =
+            AnalyticsConstants.reportablePrefKeys.mapTo(
+                HashSet(AnalyticsConstants.reportablePrefKeys.size),
+            ) { context.getString(it) }
+    }
+
+    var isEnabled: Boolean
+        get() = optIn
+        set(value) {
+            optIn = value
+            AnkiDroidApp.instance.sharedPrefs().edit {
+                putBoolean(ANALYTICS_OPTIN_KEY, value)
+            }
+            // Rebuild the underlying client so its own `enabled` flag picks
+            // up the new opt-in state without waiting for the next launch.
+            if (::appContext.isInitialized) {
+                reinitialize(appContext)
+            }
+        }
+
+    /**
+     * Switches analytics into "development" mode: forces the [samplePercentage]
+     * to 100 so every event is sent (production samples a subset to limit
+     * volume) and reinitialized the underlying analytics client to apply it.
+     *
+     * Intended for debug builds and testing not for production use.
+     */
+    fun setDevMode(context: Context) {
+        Timber.d("setDevMode() re-configuring for development analytics tagging")
+        samplePercentage = AnalyticsSamplePercentage.Full
+        reinitialize(context)
+    }
+}

--- a/common/src/main/java/com/ichi2/anki/common/utils/ext/Throwable.kt
+++ b/common/src/main/java/com/ichi2/anki/common/utils/ext/Throwable.kt
@@ -1,0 +1,18 @@
+// SPDX-FileCopyrightText: 2026 Ashish Yadav <mailtoashish693@gmail.com>
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+package com.ichi2.anki.common.utils.ext
+
+/**
+ * Walks the cause chain and returns the deepest non-null cause.
+ *
+ * Returns the receiver itself if it has no cause, and is safe against
+ * self-referential cause cycles.
+ */
+fun Throwable.getRootCause(): Throwable {
+    var result = this
+    while (result.cause != null && result.cause !== result) {
+        result = result.cause!!
+    }
+    return result
+}

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -77,6 +77,7 @@ coroutines = '1.10.2'
 desugar-jdk-libs-nio = "2.1.5"
 drawer = "1.0.3"
 espresso = '3.7.0'
+googleAnalyticsKt = "1.2.1"
 mikehardyGoogleAnalyticsJava7 = "2.0.13"
 hamcrest = "3.0"
 imageCropper = "4.7.0"
@@ -149,6 +150,7 @@ androidx-sqlite-framework = { module = "androidx.sqlite:sqlite-framework", versi
 androidx-recyclerview = { module = "androidx.recyclerview:recyclerview", version.ref = "androidxRecyclerview" }
 auto-service = { module = "com.google.auto.service:auto-service", version.ref = "autoService" }
 auto-service-annotations = { module = "com.google.auto.service:auto-service-annotations", version.ref = "autoServiceAnnotations" }
+google-analytics-kt = { module = "io.github.criticalay:google-analytics-kt", version.ref = "googleAnalyticsKt" }
 jetbrains-annotations = { module = "org.jetbrains:annotations", version.ref = "jetbrainsAnnotations" }
 skydoves-colorpickerview = { module = "com.github.skydoves:colorpickerview", version.ref = "colorPickerView" }
 commons-io = { module = "commons-io:commons-io", version.ref = "commonsIo" }

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -79,6 +79,8 @@ coroutines = '1.10.2'
 desugar-jdk-libs-nio = "2.1.5"
 drawer = "1.0.3"
 espresso = '3.7.0'
+# https://github.com/criticalAY/google-analytics-kt
+googleAnalyticsKt = "1.2.1"
 mikehardyGoogleAnalyticsJava7 = "2.0.13"
 hamcrest = "3.0"
 imageCropper = "4.7.0"
@@ -152,6 +154,7 @@ androidx-sqlite-framework = { module = "androidx.sqlite:sqlite-framework", versi
 androidx-recyclerview = { module = "androidx.recyclerview:recyclerview", version.ref = "androidxRecyclerview" }
 auto-service = { module = "com.google.auto.service:auto-service", version.ref = "autoService" }
 auto-service-annotations = { module = "com.google.auto.service:auto-service-annotations", version.ref = "autoServiceAnnotations" }
+google-analytics-kt = { module = "io.github.criticalay:google-analytics-kt", version.ref = "googleAnalyticsKt" }
 jetbrains-annotations = { module = "org.jetbrains:annotations", version.ref = "jetbrainsAnnotations" }
 skydoves-colorpickerview = { module = "com.github.skydoves:colorpickerview", version.ref = "colorPickerView" }
 commons-io = { module = "commons-io:commons-io", version.ref = "commonsIo" }

